src/self_driving/pilot/mcst/tree_node.cpp

```diff
@@ -0,0 +1,133 @@
+
+#include <random>
+#include <cmath>
+
+#include "self_driving/pilot/mcst/tree_node.h"
+#include "self_driving/pilot/action/abstract_action.h"
+#include "self_driving/pilot/pilot.h"
+#include "self_driving/pilot_util.h"
+#include "self_driving/forecast/workload_forecast.h"
+
+#define EPSILON 1e-3
+
+namespace noisepage::selfdriving::pilot {
+
+TreeNode::TreeNode(common::ManagedPointer<TreeNode> parent, action_id_t current_action,
+                   const uint64_t current_segment_cost, uint64_t later_segments_cost)
+  : is_leaf_{true}, depth_(parent == nullptr ? 0 : parent->depth_ + 1), current_action_(current_action),
+      ancestor_cost_(current_segment_cost + (parent == nullptr ? 0 : parent->ancestor_cost_)),
+      parent_(parent), number_of_visits_{1} {
+
+  if (parent != nullptr)
+    parent->is_leaf_ = false;
+  cost_ = ancestor_cost_ + later_segments_cost;
+}
+
+action_id_t TreeNode::BestSubtree(common::ManagedPointer<TreeNode> root) {
+  // Get child of least cost
+  NOISEPAGE_ASSERT(root->children_.size() > 0, "calling best child method on non-expanded nodes");
+  auto best_child = common::ManagedPointer(root->children_[0]);
+  for (auto &child : root->children_)
+    if (child->cost_ < best_child->cost_)
+      best_child = common::ManagedPointer(child);
+  return best_child->current_action_;
+}
+
+void TreeNode::UpdateCostAndVisits(uint64_t num_expansion, uint64_t leaf_cost, uint64_t new_cost) {
+  // compute cost as average of its children weighted by number of visits
+  auto new_num_visits = num_expansion - 1 + number_of_visits_;
+  cost_ = (number_of_visits_ / new_num_visits) * cost_
+          - leaf_cost / new_num_visits
+          + (num_expansion / new_num_visits) * new_cost;
+  number_of_visits_ = new_num_visits;
+}
+
+common::ManagedPointer<TreeNode> TreeNode::SampleChild() {
+  // compute max of children's cost
+  uint64_t highest = 0;
+  for (auto &child : children_)
+    highest = std::max(child->cost_, highest);
+
+  // sample based on cost and num of visits of children
+  std::vector<common::ManagedPointer<TreeNode>> selected_children, out;
+  uint64_t best_value = 0;
+  for (auto &child : children_) {
+    auto child_obj = std::pow((highest + EPSILON) / (child->cost_ + EPSILON), 2) +
+                     std::sqrt(2 * std::log(number_of_visits_) / child->number_of_visits_ );
+    if (child_obj > best_value) {
```

Comment:
Oh forgot to mention this one, which is actually important:  we shouldn't directly pick the child with the highest objective. Instead, we should really *sample* from all the children using their objectives as the probability distribution. This way we don't need the tiebreaker either.